### PR TITLE
Improvements to HTTP web proxy

### PR DESCRIPTION
**Pull request by iclaxton**
The current http proxy implementation has a few issues that I encountered when trying to implement behind a Squid proxy so we were unable to use it. This update should resolve the issues we encountered

Bug: proxy connect packet is malformed due to a space after HTTP/1.1.
Feature: Pass the host name so proxy host acls can be matched.
Feature: Switch to the WebRequest.DefaultWebProxy which allows for overriding in app.settings file and falls back to the system proxy if not configured.
Improvement: Use ordinal comparision for IndexOf on the proxy response.

### CODE REVIEW

**Comment by gbirchmeier** [comment]
Hi @IanLeeClaxton, is this still a "living" fix for you?  If I rebase to latest, can you review and help me see it through to merge?

**Comment by IanLeeClaxton** [comment]
Sure Grant, as far as I'm aware this is still being used at the company I wrote it for.

Let me know what you need me to do

**Comment by IanLeeClaxton** [comment]
@gbirchmeier, I fixed the merge conflicts and this branch is ready for you to test/merge. 

I recommend regression testing my changes again with a corporate class proxy and you could also automate testing using squid as part of your release pipeline.

**Comment by gbirchmeier** [comment]
@IanLeeClaxton any chance you can tag or pass on a contact to someone at the company that's using it?  If nothing else, I'd just like to let them know that this fix is going in.

**Comment by IanLeeClaxton** [comment]
I have sent a DM to Vijay who works there and will ask him to confirm on this thread

**Comment by vijayn1** [comment]
@gbirchmeier - please proceed with this request. Thank you @IanLeeClaxton

File: QuickFIXn/Transport/StreamFactory.cs [hunk 1]
```diff
@@ -79,7 +81,7 @@ public static Stream CreateClientStream(IPEndPoint endpoint, SocketSettings sett
             if (!settings.SocketIgnoreProxy)
             {
                 // If system has configured a proxy for this config, use it.
-                socket = CreateTunnelThruProxy(endpoint.Address.ToString(), endpoint.Port);
+                socket = CreateTunnelThruProxy(endpoint.Address.ToString(), endpoint.Port, settings.SocketConnectHost);
```
Comment (gbirchmeier) [inline QuickFIXn/Transport/StreamFactory.cs:84]: @IanLeeClaxton Why can't we use settings.SocketCommonName here?  Why did you need to create a new var SocketConnectHost?

**Comment by IanLeeClaxton** [comment]
@gbirchmeier I have applied your recommendations to this PR

File: QuickFIXn/Transport/StreamFactory.cs [hunk 2]
```diff
@@ -50,7 +50,9 @@ public static class StreamFactory
             Socket socketThruProxy = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
             socketThruProxy.Connect(proxyEndPoint);
 
-            string proxyMsg = $"CONNECT {destIp}:{destPort} HTTP/1.1 \n\n";
+            string proxyMsg = !string.IsNullOrWhiteSpace(destHostName)
```
[the remark this answers is not in the document]
Comment (gbirchmeier) [reply QuickFIXn/Transport/StreamFactory.cs]: I see you made the change, thanks.

But you didn't answer the question--  Will the new string will work with non-Squid users?  (I assume it will, just covering my bases)

**Comment by VAllens** [comment]
This PR seems to have no problem.
Let's keep an eye on it !

**Comment by IanLeeClaxton** [comment]
Probably the correct way to resolve this issue is by using a proxy bypass which can include wildcards for subdomains etc..

File: QuickFIXn/Transport/StreamFactory.cs [hunk 3]
```diff
@@ -50,7 +50,7 @@ public static class StreamFactory
             Socket socketThruProxy = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
             socketThruProxy.Connect(proxyEndPoint);
 
-            string proxyMsg = $"CONNECT {destIp}:{destPort} HTTP/1.1 \n\n";
+            string proxyMsg = $"CONNECT {destHostName}:{destPort} HTTP/1.1\nHost: {destHostName}:{destPort}\n\n";
```
[the remark this answers is not in the document]
Comment (gbirchmeier) [reply QuickFIXn/Transport/StreamFactory.cs:53]: hey?

**Comment by gbirchmeier** [comment]
Does anyone have any input on my code question?

**Comment by VAllens** [comment]
Hi, @gbirchmeier 

I think this is to be more compliant with the `HTTP/1.1` specification.
I searched for information and here is the answer to why the “Host” header was added:

When using the `HTTP/1.1` **CONNECT** method to establish a tunnel through a proxy, the most common and standard approach is to include the **target hostname and port** in the `CONNECT` request line, and also include a `Host` header. This not only adheres more closely to the `HTTP/1.1` specification (in which the `Host` header is mandatory), but is also more compatible with scenarios that require domain-based connection validation, SNI (Server Name Indication), or other proxy rules.

1. **Changing from `destIp` to `destHostName`**  
   - Some proxies rely on the hostname for forwarding or configuration (e.g., domain-based rules).  
   - Certain HTTPS requests depend on SNI (Server Name Indication), which typically uses a hostname rather than an IP address.

2. **Adding the `Host: {destHostName}:{destPort}` header**  
   - Under `HTTP/1.1`, the `Host` header is mandatory; many proxies or servers will consider requests invalid without it.  
   - Some proxies verify the `Host` header to decide if and how to forward the request.

**In summary**, these changes ensure that proxy tunnel requests conform more closely to the `HTTP/1.1` specification and work better in HTTPS or similar scenarios, reducing potential connection failures or compatibility issues caused by omitting the `Host` header or using an IP address.

**Comment by gbirchmeier** [comment]
Thanks, @VAllens, that's good enough for me!
